src/api/ucc.h

```diff
@@ -0,0 +1,709 @@
+    UCC_OP_LAND             = UCC_BIT(8),
+    UCC_OP_LOR              = UCC_BIT(9),
+    UCC_OP_LXOR             = UCC_BIT(10),
+    UCC_OP_BAND             = UCC_BIT(11),
+    UCC_OP_BOR              = UCC_BIT(12),
+    UCC_OP_BXOR             = UCC_BIT(13),
+    UCC_OP_MAXLOC           = UCC_BIT(14),
+    UCC_OP_MINLOC           = UCC_BIT(15)
+} ucc_reduction_op_t;
+
+typedef enum {
+    UCC_BARRIER            = UCC_BIT(0),
+    UCC_BCAST              = UCC_BIT(1),
+    UCC_ALLREDUCE          = UCC_BIT(2),
+    UCC_REDUCE             = UCC_BIT(3),
+    UCC_ALLTOALL           = UCC_BIT(4),
+    UCC_ALLGATHER          = UCC_BIT(5),
+    UCC_GATHER             = UCC_BIT(6),
+    UCC_SCATTER            = UCC_BIT(7),
+    UCC_MCAST              = UCC_BIT(8),
+    UCC_FANIN              = UCC_BIT(9),
+    UCC_FANOUT             = UCC_BIT(10),
+    UCC_COLL_LAST          = UCC_BIT(11)
+} ucc_coll_type_t;
+
+typedef enum {
+    UCC_THREAD_SINGLE       = UCC_BIT(0),
+    UCC_THREAD_FUNNELED     = UCC_BIT(1),
+    UCC_THREAD_MULTIPLE     = UCC_BIT(2)
+} ucc_thread_mode_t;
+
+typedef enum {
+	UCC_NO_SYNC_COLLECTIVES = 0,
+	UCC_SYNC_COLLECTIVES = 1
+} ucc_coll_sync_type_t;
+
+/**
+ * @brief UCC library initializatoin parameters
+ */
+
+typedef struct ucc_lib_params {
+    uint64_t                mask;
+    ucc_thread_mode_t       thread_mode;
+    ucc_coll_type_t         coll_types;
+    ucc_reduction_op_t      reduction_types;
+    ucc_coll_sync_type_t    sync_type;
+} ucc_lib_params_t;
+
+
+typedef struct ucc_lib_attribs {
+    uint64_t                mask;
+    ucc_thread_mode_t       thread_mode;
+    ucc_coll_type_t         coll_types;
+    ucc_reduction_op_t      reduction_types;
+    ucc_coll_sync_type_t    sync_type;
+} ucc_lib_attribs_t;
+
+
+/**
+ * @ingroup UCC_LIB
+ * @brief @ref ucc_config_read allocates the @ref ucc_lib_config_t structure and
+ * fetches the configuration values from the run-time environment. The run-time
+ * environment supported are environment variables or a configuration file.
+ *
+ *   @param [out] config_p      Pointer to configuration descriptor as defined by
+ *                              @ref ucc_lib_config_t.
+ *   @param [out] env_prefix    If not NULL, the routine searches for the
+ *                              environment variables with the prefix UCC_<env_prefix>. Otherwise, the
+ *                              routines search for the environment variables that start with the prefix @
+ *                              UCC_.
+ *   @param [in]  filename      If not NULL, read configuration values from the
+ *                              file defined by @e filename. If the file does not exist, it will be ignored
+ *                              and no error will be reported to the user.
+ *
+ * @return Error code as defined by @ref ucc_status_t
+ */
+
+ucc_status_t ucc_lib_config_read(const char *env_prefix, const char *filename,
+                                 ucc_lib_config_t **config_p);
```

Comment:
ucc_lib_config_t **config_p should be imho ucc_lib_config_h *configh_p. Since ucc_lib_config internals are never exposed to the user we should declare it as "handle", ie typedef struct ucc_lib_config* ucc_lib_config_h . ?

Reply:
Sure. I'm fine with it.